lib/THC/THCTensor.c

```diff
@@ -599,7 +599,7 @@ void THCudaTensor_freeCopyTo(THCState *state, THCudaTensor *self, THCudaTensor *
 static void THCudaTensor_rawInit(THCState *state, THCudaTensor *self)
 {
   self->refcount = 1;
-  self->storage = NULL;
+  self->storage = THCudaStorage_new(state);
```

Comment:
To address your higher-level point, I think that a device really should be assigned to a storage, because multiple tensors may be assigned to the same storage and should therefore have the same device automatically without trying to maintain that invariant manually.

However, the semantics have shifted a bit since I did this implementation - perhaps an unallocated tensor should have no storage, but once I call t:setDevice(i), it gets an empty storage with a device. Maybe then I wouldn't need this THC_DEVICE_NONE indicator, because every storage would have a real device.